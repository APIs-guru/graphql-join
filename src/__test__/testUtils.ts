declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionArgs,
  ExecutionResult,

  graphql,
  parse,
  validate,
  execute,
  print,
  buildSchema,
  printSchema,
  formatError,
  specifiedRules,
} from 'graphql';
import * as _ from 'lodash';

import { GraphQLJoinSchema, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, args: object, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    let result = `${schemaName}::${parentType.name}::${fieldName}`;
    if (Object.keys(args).length !== 0) {
      result += `;args=${JSON.stringify(args)}`
    }
    return result;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode, variableValues?: object) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    // FIXME: Should use executeQuery but blocked by
    // https://github.com/facebook/jest/issues/3917
    return graphql({
      schema,
      source: new Source(query, 'Send to ' + schemaName),
      variableValues,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string | { idl: string, prefix?: string };
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinIDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (schemaSource, name) => {
    if (typeof schemaSource === 'string') {
      schemaSource = { idl: schemaSource };
    }
    const { prefix, idl } = schemaSource;
    const schema = buildSchema(new Source(idl, name));
    stubSchema(schema);
    return { schema , prefix};
  });

  const joinSchema = new GraphQLJoinSchema(
    new Source(joinIDL, 'Join SDL'),
    remoteSchemas
  );
  const schema = joinSchema.schema;

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string | { query: string, variableValues: object },
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (queryAST) => {
          const query = print(queryAST);
          expect(query).toMatchSnapshot();
          return Promise.resolve(results[name])
        };
      }
      return makeProxy(name, schema)
    });

    const queryObj = typeof query === 'string' ?
      { query: query, variableValues: {} } : query;
    const result = await executeQuery({
      schema,
      document: parse(new Source(queryObj.query, 'ClientQuery')),
      variableValues: queryObj.variableValues,
      contextValue: new ProxyContext(joinSchema, proxyFns),
    });

    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function executeQuery(args: ExecutionArgs) {
  const validationErrors = validate(args.schema, args.document, specifiedRules);
  expect(validationErrors.join('\n')).toBe('');
  return execute(args);
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

