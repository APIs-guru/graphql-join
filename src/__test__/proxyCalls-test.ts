import { testJoin } from './testUtils';

describe('proxy calls', () => {

  test('pass random JSON as argument', async () => {
    const execute = testJoin({
      test: `
        scalar JSON
        type Query { foo(json: JSON): String }
      `,
    }, `
      type Query {
        foo(json: JSON): String @resolveWith(query: "foo")
      }
      query foo($json: JSON) @send(to: "test") {
        foo(json: $json)
      }
    `);
    await execute(`
      {
        string: foo(json: "testString")
        number: foo(json: 3.14)
        boolean: foo(json: true)
        null: foo(json: null)
        array: foo(json: [1, 2, 3, 4, 5])
        object: foo(json: { a: 1, b: 2, c: 3 })
        arrayOfObjects: foo(json: [{ a: 1 }, { b: 2 }, { c: 3}])
      }
    `);
  });

  test('pass Input object as argument', async () => {
    const execute = testJoin({
      test: `
        input Values {
          int: Int
          float: Float
          string: String
          boolean: Boolean
          arrayOfInt: [Int]
          otherValue: Values
          arrayOfValues: [Values]
        }
        type Query { foo(values: Values): String }
      `,
    }, `
      type Query {
        foo(values: Values): String @resolveWith(query: "foo")
      }
      query foo($values: Values) @send(to: "test") {
        foo(values: $values)
      }
    `);
    await execute(`
      {
        string: foo(values: { string: "testString" })
        int: foo(values: { int: 1 })
        float: foo(values: { float: 3.14 })
        boolean: foo(values: { boolean: true })
        null: foo(values: { int: null })
        arrayOfInt: foo(values: { arrayOfInt: [1, 2, 3, 4, 5] })
        otherValue: foo(values: { otherValue: { int: 1 } })
        arrayOfValues: foo(values: {
          arrayOfValues: [{ int: 1 }, { float: 2.1}]
        })
      }
    `);
  });

  test('omitting optional arguments of query', async () => {
    const execute = testJoin({
      test: `
        type Query { foo(testArg: String): String }
      `,
    }, `
      type Query {
        foo(testArg: String): String @resolveWith(query: "foo")
      }

      query foo($testVal: String) @send(to: "test") {
        foo(testArg: $testVal)
      }
    `);
    await execute('{ foo }');
  });

  test('default arguments of query', async () => {
    const execute = testJoin({
      test: `
        type Query { foo(testArg: String): String }
      `,
    }, `
      type Query {
        foo(testArg: String): String @resolveWith(query: "foo")
      }

      query foo($testArg: String = "default") @send(to: "test") {
        foo(testArg: $testArg)
      }
    `);
    await execute('{ foo }');
  });

  test('use variable inside input object', async () => {
    const execute = testJoin({
      test: `
        input TestInput {
          bar: String
          baz: String
        }
        type Query { foo(values: TestInput): String }
      `,
    }, `
      type Query {
        foo(baz: String): String @resolveWith(query: "foo")
      }
      query foo($baz: String) @send(to: "test") {
        foo(values: { bar: "FixedBarValue", baz: $baz})
      }
    `);
    await execute(`{ foo(baz: "BazValue") }`);
    await execute(`{ foo }`);
  });
});
