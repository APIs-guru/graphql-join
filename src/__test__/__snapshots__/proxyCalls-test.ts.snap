// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxy calls default arguments of query 1`] = `
type Query {
  foo(testArg: String): String
}
`;

exports[`proxy calls default arguments of query 2`] = `
{
  foo(testArg: "default")
}
`;

exports[`proxy calls default arguments of query 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test::Query::foo;args={\\"testArg\\":\\"default\\"}"
    },
    "errors": []
  }
]
`;

exports[`proxy calls omiting optional argument 1`] = `
type Query {
  foo(testArg: String): String
}
`;

exports[`proxy calls omiting optional argument 2`] = `
{
  foo
}
`;

exports[`proxy calls omiting optional argument 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test::Query::foo"
    },
    "errors": []
  }
]
`;

exports[`proxy calls omitting optional arguments of query 1`] = `
type Query {
  foo(testArg: String): String
}
`;

exports[`proxy calls omitting optional arguments of query 2`] = `
{
  foo
}
`;

exports[`proxy calls omitting optional arguments of query 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test::Query::foo"
    },
    "errors": []
  }
]
`;

exports[`proxy calls pass Input object as argument 1`] = `
type Query {
  foo(values: Values): String
}

input Values {
  int: Int
  float: Float
  string: String
  boolean: Boolean
  arrayOfInt: [Int]
  otherValue: Values
  arrayOfValues: [Values]
}
`;

exports[`proxy calls pass Input object as argument 2`] = `
{
  foo(values: {string: "testString"})
}
`;

exports[`proxy calls pass Input object as argument 3`] = `
{
  foo(values: {int: 1})
}
`;

exports[`proxy calls pass Input object as argument 4`] = `
{
  foo(values: {float: 3.14})
}
`;

exports[`proxy calls pass Input object as argument 5`] = `
{
  foo(values: {boolean: true})
}
`;

exports[`proxy calls pass Input object as argument 6`] = `
{
  foo(values: {int: null})
}
`;

exports[`proxy calls pass Input object as argument 7`] = `
{
  foo(values: {arrayOfInt: [1, 2, 3, 4, 5]})
}
`;

exports[`proxy calls pass Input object as argument 8`] = `
{
  foo(values: {otherValue: {int: 1}})
}
`;

exports[`proxy calls pass Input object as argument 9`] = `
{
  foo(values: {arrayOfValues: [{int: 1}, {float: 2.1}]})
}
`;

exports[`proxy calls pass Input object as argument 10`] = `
[
  "\\n      {\\n        string: foo(values: { string: \\"testString\\" })\\n        int: foo(values: { int: 1 })\\n        float: foo(values: { float: 3.14 })\\n        boolean: foo(values: { boolean: true })\\n        null: foo(values: { int: null })\\n        arrayOfInt: foo(values: { arrayOfInt: [1, 2, 3, 4, 5] })\\n        otherValue: foo(values: { otherValue: { int: 1 } })\\n        arrayOfValues: foo(values: {\\n          arrayOfValues: [{ int: 1 }, { float: 2.1}]\\n        })\\n      }\\n    ",
  {
    "data": {
      "string": "test::Query::foo;args={\\"values\\":{\\"string\\":\\"testString\\"}}",
      "int": "test::Query::foo;args={\\"values\\":{\\"int\\":1}}",
      "float": "test::Query::foo;args={\\"values\\":{\\"float\\":3.14}}",
      "boolean": "test::Query::foo;args={\\"values\\":{\\"boolean\\":true}}",
      "null": "test::Query::foo;args={\\"values\\":{\\"int\\":null}}",
      "arrayOfInt": "test::Query::foo;args={\\"values\\":{\\"arrayOfInt\\":[1,2,3,4,5]}}",
      "otherValue": "test::Query::foo;args={\\"values\\":{\\"otherValue\\":{\\"int\\":1}}}",
      "arrayOfValues": "test::Query::foo;args={\\"values\\":{\\"arrayOfValues\\":[{\\"int\\":1},{\\"float\\":2.1}]}}"
    },
    "errors": []
  }
]
`;

exports[`proxy calls pass random JSON as argument 1`] = `
scalar JSON

type Query {
  foo(json: JSON): String
}
`;

exports[`proxy calls pass random JSON as argument 2`] = `
{
  foo(json: "testString")
}
`;

exports[`proxy calls pass random JSON as argument 3`] = `
{
  foo(json: 3.14)
}
`;

exports[`proxy calls pass random JSON as argument 4`] = `
{
  foo(json: true)
}
`;

exports[`proxy calls pass random JSON as argument 5`] = `
{
  foo(json: null)
}
`;

exports[`proxy calls pass random JSON as argument 6`] = `
{
  foo(json: [1, 2, 3, 4, 5])
}
`;

exports[`proxy calls pass random JSON as argument 7`] = `
{
  foo(json: {a: 1, b: 2, c: 3})
}
`;

exports[`proxy calls pass random JSON as argument 8`] = `
{
  foo(json: [{a: 1}, {b: 2}, {c: 3}])
}
`;

exports[`proxy calls pass random JSON as argument 9`] = `
[
  "\\n      {\\n        string: foo(json: \\"testString\\")\\n        number: foo(json: 3.14)\\n        boolean: foo(json: true)\\n        null: foo(json: null)\\n        array: foo(json: [1, 2, 3, 4, 5])\\n        object: foo(json: { a: 1, b: 2, c: 3 })\\n        arrayOfObjects: foo(json: [{ a: 1 }, { b: 2 }, { c: 3}])\\n      }\\n    ",
  {
    "data": {
      "string": "test::Query::foo;args={\\"json\\":\\"testString\\"}",
      "number": "test::Query::foo;args={\\"json\\":3.14}",
      "boolean": "test::Query::foo;args={\\"json\\":true}",
      "null": "test::Query::foo;args={\\"json\\":null}",
      "array": "test::Query::foo;args={\\"json\\":[1,2,3,4,5]}",
      "object": "test::Query::foo;args={\\"json\\":{\\"a\\":1,\\"b\\":2,\\"c\\":3}}",
      "arrayOfObjects": "test::Query::foo;args={\\"json\\":[{\\"a\\":1},{\\"b\\":2},{\\"c\\":3}]}"
    },
    "errors": []
  }
]
`;

exports[`proxy calls use variable inside input object 1`] = `
type Query {
  foo(baz: String): String
}
`;

exports[`proxy calls use variable inside input object 2`] = `
{
  foo(values: {bar: "FixedBarValue", baz: "BazValue"})
}
`;

exports[`proxy calls use variable inside input object 3`] = `
[
  "{ foo(baz: \\"BazValue\\") }",
  {
    "data": {
      "foo": "test::Query::foo;args={\\"values\\":{\\"bar\\":\\"FixedBarValue\\",\\"baz\\":\\"BazValue\\"}}"
    },
    "errors": []
  }
]
`;

exports[`proxy calls use variable inside input object 4`] = `
{
  foo(values: {bar: "FixedBarValue"})
}
`;

exports[`proxy calls use variable inside input object 5`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test::Query::foo;args={\\"values\\":{\\"bar\\":\\"FixedBarValue\\"}}"
    },
    "errors": []
  }
]
`;
